homeassistant/components/ecovacs/__init__.py

```diff
@@ -54,56 +42,19 @@ async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
 
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up this integration using UI."""
+    controller = EcovacsController(hass, entry.data)
+    await controller.initialize()
 
-    def get_devices() -> list[VacBot]:
-        ecovacs_api = EcoVacsAPI(
-            get_client_device_id(),
-            entry.data[CONF_USERNAME],
-            EcoVacsAPI.md5(entry.data[CONF_PASSWORD]),
-            entry.data[CONF_COUNTRY],
-            entry.data[CONF_CONTINENT],
-        )
-        ecovacs_devices = ecovacs_api.devices()
-
-        _LOGGER.debug("Ecobot devices: %s", ecovacs_devices)
-        devices: list[VacBot] = []
-        for device in ecovacs_devices:
-            _LOGGER.debug(
-                "Discovered Ecovacs device on account: %s with nickname %s",
-                device.get("did"),
-                device.get("nick"),
-            )
-            vacbot = VacBot(
-                ecovacs_api.uid,
-                ecovacs_api.REALM,
-                ecovacs_api.resource,
-                ecovacs_api.user_access_token,
-                device,
-                entry.data[CONF_CONTINENT],
-                monitor=True,
-            )
-
-            devices.append(vacbot)
-        return devices
-
-    hass.data.setdefault(DOMAIN, {})[
-        entry.entry_id
-    ] = await hass.async_add_executor_job(get_devices)
-
-    async def async_stop(event: object) -> None:
-        """Shut down open connections to Ecovacs XMPP server."""
-        devices: list[VacBot] = hass.data[DOMAIN][entry.entry_id]
-        for device in devices:
-            _LOGGER.info(
-                "Shutting down connection to Ecovacs device %s",
-                device.vacuum.get("did"),
-            )
-            await hass.async_add_executor_job(device.disconnect)
-
-    # Listen for HA stop to disconnect.
-    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, async_stop)
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = controller
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+    return True
 
-    if hass.data[DOMAIN][entry.entry_id]:
-        await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
 
-    return True
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload config entry."""
+    if unload_ok := await hass.config_entries.async_unload_platforms(entry, PLATFORMS):
+        await hass.data[DOMAIN][entry.entry_id].teardown()
+        del hass.data[DOMAIN][entry.entry_id]
+        if len(hass.data[DOMAIN]) == 0:
+            hass.data.pop(DOMAIN)
```

Comment:
We don't really do this anywhere else

Reply:
That's not true. `abode` and `adguard` is doing it to mention the first two